### PR TITLE
Replace DataCache.Find by DataCache.Seek

### PR DESCRIPTION
**Pull request by Tommo-L**
**Background**

The implementation of economic model needs to use db.Seek method.

**Where in the software does this update applies to?**
- Persistence(leveldb/rocksdb)

**Where in the software does this update applies to?**
- Persistence
- Modules

**Checklist**

- [x] Did you add tests covering your changes?
- [x] Did you create the related issue or pr for the affected modules.

### CODE REVIEW

File: src/neo/IO/Caching/DataCache.cs [hunk 1]
```diff
@@ -201,7 +201,29 @@ public IEnumerable<(TKey Key, TValue Value)> Find(byte[] key_prefix = null)
...
+        /// <returns>Entries found with the desired prefix</returns>
+        public IEnumerable<(TKey Key, TValue Value)> Find(byte[] key_prefix = null)
+        {
+            var enumerator = Seek(key_prefix, SeekDirection.Forward).GetEnumerator();
+            while (enumerator.MoveNext())
+                if (enumerator.Current.Key.ToArray().AsSpan().StartsWith(key_prefix))
+                    yield return enumerator.Current;
+        }
+
+        public IEnumerable<(TKey Key, TValue Value)> FindRange(TKey start, TKey end)
```
Comment (Tommo-L) [inline src/neo/IO/Caching/DataCache.cs:217]: Should I remove this method first? But I want to use it in #1716

File: src/neo/IO/Caching/DataCache.cs [hunk 2]
```diff
@@ -149,25 +149,26 @@ public void Delete(TKey key)
         /// </summary>
         /// <param name="key_prefix">Must maintain the deserialized format of TKey</param>
         /// <returns>Entries found with the desired prefix</returns>
-        public IEnumerable<(TKey Key, TValue Value)> Find(byte[] key_prefix = null)
+        public IEnumerable<(TKey Key, TValue Value)> Seek(byte[] keyOrPrefix = null, SeekDirection direction = SeekDirection.Forward)
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:152]: The comments should be updated too.

File: src/neo/IO/Caching/DataCache.cs [hunk 3]
```diff
@@ -145,9 +145,10 @@ public void Delete(TKey key)
         protected abstract void DeleteInternal(TKey key);
 
         /// <summary>
-        /// Find the entries that start with the `key_prefix`
+        /// Find the entries that start with the `keyOrPrefix`
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:148]: I think this is incorrect.

File: src/neo/IO/Caching/DataCache.cs [hunk 4]
```diff
@@ -145,11 +145,11 @@ public void Delete(TKey key)
         protected abstract void DeleteInternal(TKey key);
 
         /// <summary>
-        /// Find the entries that start with the `keyOrPrefix`
+        /// Find the entries that not less/more than the `keyOrPrefix`
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:148]:
```suggestion
        /// Seek to the entry with the specific key
```

File: src/neo/IO/Caching/DataCache.cs [hunk 5]
```diff
@@ -145,11 +145,11 @@ public void Delete(TKey key)
...
         /// </summary>
         /// <param name="keyOrPrefix">Must maintain the deserialized format of TKey</param>
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:150]:
```suggestion
        /// <param name="keyOrPrefix">The key to be sought</param>
```

File: src/neo/IO/Caching/DataCache.cs [hunk 6]
```diff
@@ -145,11 +145,11 @@ public void Delete(TKey key)
...
-        /// <param name="direction">Forward or Backward</param>
-        /// <returns>Entries found with the desired prefix</returns>
+        /// <param name="direction">Forward means not less than, Backward means not more than</param>
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:151]:
```suggestion
        /// <param name="direction">The direction of the seeking.</param>
```

File: src/neo/IO/Caching/DataCache.cs [hunk 7]
```diff
@@ -145,11 +145,11 @@ public void Delete(TKey key)
...
+        /// <returns>Entries that are not less/more than the `keyOrPrefix`</returns>
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:152]:
```suggestion
        /// <returns>An enumerator containing all the entries after seeking.</returns>
```

File: src/neo/IO/Caching/DataCache.cs [hunk 8]
```diff
@@ -145,29 +145,31 @@ public void Delete(TKey key)
...
+        public IEnumerable<(TKey Key, TValue Value)> Seek(byte[] keyOrPrefix = null, SeekDirection direction = SeekDirection.Forward)
         {
             IEnumerable<(byte[], TKey, TValue)> cached;
             HashSet<TKey> cachedKeySet;
+            ByteArrayComparer comparer = direction == SeekDirection.Forward ? ByteArrayComparer.Default : ByteArrayComparer.Reverse;
             lock (dictionary)
             {
                 cached = dictionary
-                    .Where(p => p.Value.State != TrackState.Deleted && (key_prefix == null || p.Key.ToArray().AsSpan().StartsWith(key_prefix)))
+                    .Where(p => p.Value.State != TrackState.Deleted && (keyOrPrefix == null || comparer.Compare(p.Key.ToArray(), keyOrPrefix) >= 0))
```
Comment (erikzhang) [inline src/neo/IO/Caching/DataCache.cs:161]: If keyOrPrefix is null and direction is backward, you will find nothing. But it should find everything.

**Pull request merged**